### PR TITLE
Item counters added to Communities and Collections REST API

### LINKED ISSUE #1787: Add option to display Item counts for Communities and Collections
**Is your feature request related to a problem? Please describe.**
Please replicate in DSpace 7 the DSpace 6 XMLUI option to display items counts for Communities and Collections.

**Describe the solution you'd like**
The display in DSpace 6 XMLU is fine.
DSpace6_XMLUI_item_counts
DSpace6_XMLUI_item_counts2

**Describe alternatives or workarounds you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by damian-joz**
## References
_Add references/links to any related issues or PRs. These may include:_
* Backend PR needed to fix https://github.com/DSpace/dspace-angular/issues/1787
* Frontend PR: https://github.com/DSpace/dspace-angular/pull/2250
* REST Contact: https://github.com/DSpace/RestContract/pull/231

## Description
This initial PR of item counts at Communites and Collections list. It is ported from DSpace 6.x

## Instructions for Reviewers
Please add a more detailed description of the changes made by your PR. At a minimum, providing a bulleted list of changes in your PR is helpful to reviewers.

List of changes in this PR:
* Collection and Community REST models extended with one field
* Collection and Community REST converters modified to set item counts field

**Include guidance for how to test or review your PR.** This may include: steps to reproduce a bug, screenshots or description of a new feature, or reasons behind specific changes. 

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [ x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [ x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [ ] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [ ] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [ ] If my PR includes new libraries/dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [ ] If my PR modifies REST API endpoints, I've opened a separate [REST Contract](https://github.com/DSpace/RestContract/blob/main/README.md) PR related to this change.
- [ ] If my PR includes new configurations, I've provided basic technical documentation in the PR itself.
- [ ] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

### CODE REVIEW

File: dspace-api/src/main/java/org/dspace/content/dao/impl/CollectionDAOImpl.java
Comment (damian-joz) [inline dspace-api/src/main/java/org/dspace/content/dao/impl/CollectionDAOImpl.java:185]: This will be removed as ItemCounter has own methods.

**Comment by tdonohue** [comment]
@damian-joz : This PR appears to be failing our automated checks.  Currently it's failing checkstyle checks because the code doesn't align with our Code Style: https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide   You can see the errors thrown by checking the codestyle via commandline as documented here: https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide#CodeStyleGuide-Checkingcodestylefromcommandline  (There are also IDE integrations on that same page).

Keep in mind, backend PRs should align with the checklist above, which is a shortened version of our contribution docs: https://github.com/DSpace/DSpace/blob/main/CONTRIBUTING.md

**Comment by damian-joz** [comment]
> @damian-joz : This PR appears to be failing our automated checks. Currently it's failing checkstyle checks because the code doesn't align with our Code Style: https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide You can see the errors thrown by checking the codestyle via commandline as documented here: https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide#CodeStyleGuide-Checkingcodestylefromcommandline (There are also IDE integrations on that same page).
> 
> Keep in mind, backend PRs should align with the checklist above, which is a shortened version of our contribution docs: https://github.com/DSpace/DSpace/blob/main/CONTRIBUTING.md

@tdonohue 
I've made some fixes to the PR.

**Review by kshepherd — DISMISSED** [review]
Hi @damian-joz , thanks for the PR! Looks great overall.

I've tested this along with [the dspace-angular#2250 ](https://github.com/DSpace/dspace-angular/pull/2250) and it works as advertised. I have added some inline comments where I think javadoc would be helpful to explain how the item counter works and what kind of value is being get/set in the community and collection rest models. Pretty minor changes, overall it is working well and I think very close to a +1!

I did leave a note about ItemCounter being implemented in quite an old way, and that rewriting it as a service in the modern DSpace "way" might be nice but since it's old code that works and was used in 6.x implementations already, I wouldn't personally consider that a blocker. Some comments and javadoc will help to explain the usage, though.

Also, just a note - the way the item count DAO impl works, it will count each version as a separate item. I noticed this in my own test environment. Again, this is existing code and so the same will be true for prior implementations.

Do we have RestContract PR to document the new Community and Collection fields? This will be required.

File: dspace-api/src/main/java/org/dspace/browse/ItemCounter.java
Comment (damian-joz) [reply dspace-api/src/main/java/org/dspace/browse/ItemCounter.java:71]: @kshepherd  That was exactly my concern. The old service works but it doesn't fit the current code style. I will add javadoc but as you said, that should be refactored in the future.

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/model/CollectionRest.java
Comment (damian-joz) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/model/CollectionRest.java:78]: I would change the name.

**Review by tdonohue — CHANGES_REQUESTED** [review]
@damian-joz : Thanks! I haven't had a chance to test this out myself, but wanted to give some quick feedback based on the code:

* As @kshepherd mentions, we need a corresponding PR to the REST Contract to document the changes to the Collection / Community responses: https://github.com/DSpace/RestContract
* I think we need to add a way to *disable* this feature.  In DSpace 6, this was managed via a config named `webui.strengths.show`: https://github.com/DSpace/DSpace/blob/dspace-6_x/dspace/config/dspace.cfg#L938
    * I think this is VERY important because in 6.x there were some sites which found this feature had bad performance (for very large sites).  So, we need to be able to turn if off if it is impacting the performance of the site.
    * When it's disabled, the archived items count returned in the REST API response for communities & collections could just be either missing (not included in the response) or empty ("").  Then the UI code would need to be updated to handle that scenario and display nothing if no value exists.
* Could you create some basic automated tests for this new code?   Perhaps the best place would be in `CollectionRestRepositoryIT` and `CommunityRestRepositoryIT` as those are where we run automated tests of the Community/Collection endpoints.
    * We should have basic tests just to prove that the count is included in the responses for communities & collections when enabled.  And that it's missing (or empty) when disabled.

Beyond that, the basic code looks fine.  I'll still try to find time to test this soon, but wanted to give you immediate feedback based on a code review today.  Thanks again.

**Comment by damian-joz** [comment]
@tdonohue 

> I think we need to add a way to disable this feature. In DSpace 6, this was managed via a config named 

Should it be enabled by default?

**Comment by tdonohue** [comment]
@damian-joz : Yes, I'd recommend we disable by default (if possible), since that was the behavior of v6.

**Comment by tdonohue** [comment]
@damian-joz : If you have a chance to get back to this, our deadline for features into 7.6 is this Friday.  If you cannot get to it in time, no worries, but we'll have to likely reschedule for 8.0 (due in 2024).

**Comment by damian-joz** [comment]
Hi @tdonohue 
I have made it managable as you said: https://github.com/DSpace/DSpace/pull/8843/commits/e31a2238824e79a2b60bf2cb1902352f5e3b9845

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/model/CollectionRest.java
Comment (damian-joz) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/model/CollectionRest.java]: @kshepherd I've renamed the field int the recent commit.

**Comment by damian-joz** [comment]
@kshepherd  I've added commit with your suggestions.

File: dspace-api/src/main/java/org/dspace/content/Collection.java
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/content/Collection.java:345]: Tiny thing, but this method should probably be plural `countArchivedItems` instead of `countArchivedItem`.  But, if you'd rather leave as-is, that's fine.

File: dspace-api/src/main/java/org/dspace/content/Community.java
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/content/Community.java:273]: Same here, this method should probably be plural `countArchivedItems` instead of `countArchivedItem`.

File: dspace-api/src/main/java/org/dspace/content/service/CollectionService.java
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/content/service/CollectionService.java:481]: Same here. This method should probably be plural `countArchivedItems` instead of `countArchivedItem`.

File: dspace-api/src/main/java/org/dspace/content/service/CommunityService.java
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/content/service/CommunityService.java:304]: I'd recommend renaming this `countArchivedItems()` so that it's the same name as in `CollectionService`.  Methods that do the same thing should have the same name.

File: dspace-api/src/main/java/org/dspace/browse/ItemCounter.java
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/browse/ItemCounter.java:110]: Rather than returning zero (0), I'd recommend this return `-1` if it is disabled.  This way you can actually tell if it's disabled or not.  If it returns zero, it could be that the Collection/Community is empty.

(I notice that the current behavior is that if the value is zero, it just doesn't display.  I don't think it behaved that way in 6.x.  In 6.x, if the collection was empty it showed a count of `0`.)

File: dspace/config/dspace.cfg
Comment (tdonohue) [inline dspace/config/dspace.cfg:1058]: Again, I'd recommend this be false by default.  Same as in 6.x.  Once we are able to better test performance we can turn them on by default at a later date.

**Review by tdonohue — CHANGES_REQUESTED** [review]
@damian-joz : Thanks for the updates!  This is honestly looking very good... while I'm still requesting changes, the changes I'm recommending seem minor in nature.  So, I'm hopeful you can make them relatively quickly (and minor updates to the angular PR may be needed too).

The *good news* for this PR is that our deadline for 7.6 is pushed back a week.  I'd still highly recommend trying to finish up these changes early next week if you want this in 7.6.  I think it's nearly ready to merge if you can apply these small changes.

See inline comments below.

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/converter/CollectionConverter.java
Comment (damian-joz) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/converter/CollectionConverter.java:29]: @tdonohue  As archivedItemsCount is the part of response class I am not sure if we can ignore it on demand and we should return -1 to frontend as well.

File: dspace-api/src/main/java/org/dspace/browse/ItemCounter.java
Comment (damian-joz) [reply dspace-api/src/main/java/org/dspace/browse/ItemCounter.java]: Done.

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/converter/CollectionConverter.java
Comment (damian-joz) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/converter/CollectionConverter.java]: Done.

File: dspace-api/src/main/java/org/dspace/content/service/CollectionService.java
Comment (damian-joz) [reply dspace-api/src/main/java/org/dspace/content/service/CollectionService.java]: Done.

File: dspace-api/src/main/java/org/dspace/content/Community.java
Comment (damian-joz) [reply dspace-api/src/main/java/org/dspace/content/Community.java]: Done.

File: dspace-api/src/main/java/org/dspace/content/Collection.java
Comment (damian-joz) [reply dspace-api/src/main/java/org/dspace/content/Collection.java]: Done.

File: dspace/config/dspace.cfg
Comment (damian-joz) [reply dspace/config/dspace.cfg]: Done.

**Review by tdonohue — APPROVED** [review]
👍 Thank you @damian-joz for all your hard work on this!  This looks great now and works well. Tested this with the frontend PR & I've verified that the counts only appear when enabled, and Communities/Collections with zero items now show `[0]`.

**Comment by tdonohue** [comment]
Merging for 7.6 as this has had general approval and I've verified that all my feedback & @kshepherd 's feedback has been addressed.

**Comment by alanorth** [comment]
This works well. Thank you @damian-joz. I'm wondering why the item counts use Bootstrap `lead` classes, though? This increases the font size and makes them stand out too much in my opinion.

This is my custom theme, but the effect is the same in the default theme:

<p align="center">
  
</p>

In DSpace 6 XMLUI the item counts were the same color as the default body text. If we want to have them look subtly different, we could use [Bootstrap's `font-weight-light`](https://getbootstrap.com/docs/4.6/utilities/text/#font-weight-and-italics):

<p align="center">
  
</p>

**Comment by damian-joz** [comment]
@alanorth Without that class it was to small and didn't look well. But I like your idea with bootstrap class. @tdonohue  Can we still make one small change with CSS styles as @alanorth is suggesting?

**Comment by alanorth** [comment]
Oh, I commented on the backend half of the PR instead of the Angular one...

Anyways, yes I suggest using `font-weight-light` instead of `lead` if possible, @tdonohue. Cheers!

**Comment by tdonohue** [comment]
@damian-joz : Yes, there's still time for a small PR to correct CSS.  It'd have to be a *new* PR though as this one is merged already.  The release will be finished in about one week, but small changes like this can still get in.

**Pull request merged**
